presto-main/src/main/java/com/facebook/presto/operator/aggregation/histogram/GroupedTypedHistogram.java

```diff
@@ -0,0 +1,541 @@
+            out.appendNull();
+        }
+        else {
+            BlockBuilder blockBuilder = out.beginBlockEntry();
+
+            iterateGroupNodes(currentGroupId, nodePointer -> {
+                checkArgument(nodePointer != NULL, "should never see null here as we exclude in iterateGroupNodesCall");
+                ValueNode valueNode = bucketNodeFactory.createValueNode(nodePointer);
+                valueNode.writeNodeAsRow(values, blockBuilder);
+            });
+
+            out.closeEntry();
+        }
+    }
+
+    @Override
+    public void addAll(TypedHistogram other)
+    {
+        addAll(currentGroupId, other);
+    }
+
+    @Override
+    public void readAllValues(HistogramValueReader reader)
+    {
+        iterateGroupNodes(currentGroupId, nodePointer -> {
+            checkArgument(nodePointer != NULL, "should never see null here as we exclude in iterateGroupNodesCall");
+            ValueNode valueNode = bucketNodeFactory.createValueNode(nodePointer);
+            reader.read(values, valueNode.getValuePosition(), valueNode.getCount());
+        });
+    }
+
+    @Override
+    public TypedHistogram setGroupId(long groupId)
+    {
+        this.currentGroupId = groupId;
+        return this;
+    }
+
+    @Override
+    public Type getType()
+    {
+        return type;
+    }
+
+    @Override
+    public int getExpectedSize()
+    {
+        return bucketId;
+    }
+
+    @Override
+    public boolean isEmpty()
+    {
+        return isCurrentGroupEmpty();
+    }
+
+    @Override
+    public void add(int position, Block block, long count)
+    {
+        checkState(currentGroupId != -1, "setGroupId() not called yet");
+        add(currentGroupId, block, position, count);
+    }
+
+    private void verifyCapacity()
+    {
+        if (nextNodePointer >= maxFill) {
+            rehash();
+        }
+    }
+
+    private static int computeBucketCount(int expectedSize, float mainFillRate)
+    {
+        return arraySize(expectedSize, mainFillRate);
+    }
+
+    private void addAll(long groupId, TypedHistogram other)
+    {
+        other.readAllValues((block, position, count) -> add(groupId, block, position, count));
+    }
+
+    private void add(long groupId, Block block, int position, long count)
+    {
+        verifyCapacity();
+
+        BucketDataNode bucketDataNode = bucketNodeFactory.createBucketDataNode(groupId, block, position);
+
+        if (bucketDataNode.processEntry(groupId, block, position, count)) {
+            nextNodePointer++;
+        }
+    }
+
+    private boolean isCurrentGroupEmpty()
+    {
+        return headPointers.get(currentGroupId) == NULL;
+    }
+
+    /**
+     * used to iterate over all non-null nodes in the data structure
+     *
+     * @param nodeReader - will be passed every non-null nodePointer
+     */
+    private void iterateGroupNodes(long groupdId, NodeReader nodeReader)
+    {
+        int currentPointer = (int) headPointers.get(groupdId);
```

Comment:
groupId is a long
i wasn't sure where it was safe to cast down...but i guess if here, it's ok above

Reply:
it comes into the state as a long, so i'll cast as it enters into to the histogram.

but do note we might want to check for > intMax

i don't know if the system as a whole allows > 2B groups, but i don't see why not
but
we opened ourselves up to maybe  needing long as an index into buckets (it's an int) because we now store all groups here. if we have > 2B groups, each with >= 1 value, we break all the int assumptoins

i'll cast on the way in, change currentGroupId to int

but we should reconcile this. it might mean changing all our pointers to longs 

lmk

Reply:
I have a feeling we're going to recon with this--we have  an int-bound data structure that used to live in a long-sized array. we squished everything into the original

The fundamental limitation is the single BlockBuilder and it's int interface.

we could use an array of them in ValueStore. We' have to be careful to track per-group max as when we do output, we do a single group to a BlockBuilder which is int bound


worth filing a future issue as possible problem so we aren't surprised?

OTOH, we are expecting cases of many groups with few elements, so it is not expected to be an issue?